WSRequest: Normalize URL

htmldoug opened the pull request:
# Pull Request Checklist

* [x] Have you read through the [contributor guidelines](https://www.playframework.com/contributing)?
* [x] Have you signed the [Typesafe CLA](https://www.typesafe.com/contribute/cla)?
* [x] Have you [squashed your commits](https://www.playframework.com/documentation/latest/WorkingWithGit#Squashing-commits)?
* [x] Have you added copyright headers to new files?
* [x] Have you checked that both Scala and Java APIs are updated?
* [x] Have you updated the documentation for both Scala and Java sections?
* [x] Have you added tests for any changed functionality?

## Fixes

Fixes #267

## Purpose

Ensure that `StandaloneWSRequest.uri` does not return invalid values or throw, by normalizing the `url: String` when a `StandaloneWSRequest` is constructed. Same approach is used by the java implementation.

## Background Context

This approach follows the "tolerant" recommendation, https://github.com/playframework/play-ws/issues/267#issuecomment-419266583.

Ensures:
1. `StandaloneWSRequest.url` path is encoded.
2. Any query params present in the URL are moved to `StandaloneWSRequest.queryString`.

Implementation is based upon a `WSRequestFilter`-like decorator I've been using in production since June 2018. An advantage to fixing the request right away is that `request.uri` can be invoked safely sooner in the process.

### Caveats
Normalization can still be bypassed by:
* `StandaloneAhcWSClient.copy(url = ">^..^<", ...)`
* `new StandaloneAhcWSRequest(client, ">^..^<", ...)`

htmldoug commented:
@marcospereira I've removed "WIP", and I'm ready for a review now, please.

htmldoug commented:
@gmethvin @wsargent ready for review. I see your fingerprints here from #80 and #189.

htmldoug commented:
@marcospereira @gmethvin @wsargent can I get a review please?

wsargent commented:
I'm free this weekend, I can take a look at it

wsargent commented on play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSClient.scala:
```diff
@@ -163,5 +156,106 @@ object StandaloneAhcWSClient {
+  /**
+   * Ensures:
+   * 1. [[StandaloneWSRequest.url]] path is encoded.
+   * 2. Any query params present in the URL are moved to [[StandaloneWSRequest.queryString]].
```
Provide an example URL i.e. `ws.url("http://example.com/?foo=bar")` -> `ws.url("http://example.com/").withQueryString("foo" -> "bar")`

wsargent commented on play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSClient.scala:
```diff
@@ -163,5 +156,106 @@ object StandaloneAhcWSClient {
+    try {
+      // Recover from https://github.com/AsyncHttpClient/async-http-client/issues/1149
+      Uri.create(req.url)
+      if (req.uri.getQuery == null) {
```
If `req.uri.getQuery == null` then you want to try `Uri.create(req.url)` -- otherwise you'll throw and catch an exception but will be calling `repair()` in any case.  If you can avoid throwing the exception it'll speed the happy path.

htmldoug replied:
I need some clarification to understand this comment.

> If `req.uri.getQuery == null` then you want to try `Uri.create(req.url)--`

`Uri.create(req.url)` is called on the line above, 169.

> otherwise you'll throw and catch an exception

> If you can avoid throwing the exception it'll speed the happy path.

Which exception? The one from `ahc.Uri` or `java.net.URI`? `ahc.Uri` is tolerant and should only throw if critical parts are missing, e.g. https://github.com/AsyncHttpClient/async-http-client/issues/1149. `java.net.URI` is stricter and I'd expect it throw if the query params are not encoded. I wouldn't consider either to be on the happy path since they're both avoidable by the caller.

wsargent replied:
So something like

```
 @throws[IllegalArgumentException]("if the url is unrepairable")
  private[ahc] def normalize(req: StandaloneWSRequest): StandaloneWSRequest = {
    if (req.uri.getQuery == null) {
      req
    } else {
        try {
          Uri.create(req.url)
          repair(req)
        } catch {
          case NonFatal(_) =>
            // URI parsing error
            repair(req)
        }
    }
}
```

wsargent commented on play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSClient.scala:
```diff
@@ -163,5 +156,106 @@ object StandaloneAhcWSClient {
+  }
+
+  /**
+   * Builds an AHC [[Uri]] with all parts URL encoded.
```
Can we specify the RFC here used for the encoding?

htmldoug replied:
I assume `UriEncoder.FIXING` returns an `org.asynchttpclient.uri.Uri` capable of producing an RFC 3986 compliant `String`. It makes no guarantees in the javadoc, but the private static constant names in `Utf8UrlEncoder` mention RFC3986. I'm not sure we can go so far as to claim that the `Uri` result conforms to the RFC though. The result seems closer to it at least.

I've updated the scaladoc of this method to say that it runs it through `UriEncoder.FIXING`, which is at least 100% accurate.

wsargent commented on play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSClient.scala:
```diff
@@ -163,5 +156,106 @@ object StandaloneAhcWSClient {
+    encodedQueryString: Option[String]): StandaloneWSRequest = {
+    val queryParams: List[(String, String)] = for {
+      queryString <- encodedQueryString.toList
+      // https://stackoverflow.com/a/13592567 for all of this.
```
Is this decoding the names and parameters though? 
https://stackoverflow.com/questions/13592236/parse-a-uri-string-into-name-value-collection/13592567#comment18632046_13592567

htmldoug replied:
Good catch. It's not, but I wasn't getting the encoded string from `java.net.URI` as I intended either. Fixed now and added tests.

wsargent commented on play-ahc-ws-standalone/src/test/scala/play/api/libs/ws/ahc/AhcWSRequestSpec.scala:
```diff
@@ -108,6 +107,47 @@ class AhcWSRequestSpec extends Specification with Mockito with AfterAll with Def
 
     }
 
+    "with unencoded values" in {
```
Some more tests would be good here, especially showing parameter names that are encoded.

htmldoug replied:
Added.

htmldoug commented:
@wsargent thanks for the review. I'll update with RFC links, examples, and more tests.

htmldoug commented on play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSClient.scala:
```diff
@@ -163,5 +156,106 @@ object StandaloneAhcWSClient {
+
+  /**
+   * Ensures:
+   * 1. [[StandaloneWSRequest.url]] path is encoded.
```
https://github.com/playframework/play-ws/pull/288#discussion_r249617018

htmldoug commented on play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSClient.scala:
```diff
@@ -164,5 +158,130 @@ object StandaloneAhcWSClient {
+    } yield key -> value
+
+    req
+      .withUrl(urlNoQueryParams)
```
One concern: `withUrl()` calls `normalize()` which could lead to infinite mutual recursion if I missed a case. I'd prefer to call the `StandaloneAhcWSRequest.copy()` escape hatch here, but it requires access to the implicit `Materializer` which is not accessible.

Some ideas in order of preference (least -> most):
* Leave it as is and let users report cases we missed. Should happen pretty quickly, but ugh.
* Remove the check from `withUrl()`. That'd weaken the integrity of `StandaloneAhcWSRequest`, although it's already comprisable with `copy()` if you provide your own `Materializer.`
* Change visibility to `StandaloneAhcWSRequest(...)(implicit private[ahc] val materializer: Materializer)` and do the `copy()`.

htmldoug replied:
I'm going to go ahead and push up a second commit with the last approach. If you like it, I'll squash. Otherwise, I'll drop it.

htmldoug replied:
Done. Thoughts?

htmldoug commented:
Thanks again for the review and sorry for the long delay. Holidays derailed me, but https://blog.playframework.com/play-2-7-0-rc9-released/ was a nice kick in the pants.

I've addressed all comments, and added benchmarks and more tests. Ready for another review.

htmldoug commented:
@marcospereira @wsargent do you guys need anything else from me for this to be mergeable?

htmldoug commented:
We're starting on the play 2.7 upgrade now. Any chance of getting this merged, or should we just port this to a wrapper and wrap all of our `WSClient`s to keep `WSRequest.uri` from throwing runtime exceptions?

jeffmay commented:
Any update on this? I would like to upgrade to Play 2.7 with this as well, and I have the same issue that Doug is trying to solve here.

htmldoug commented:
It's been ready to go since https://github.com/playframework/play-ws/pull/288#issuecomment-456252800. I've rebased onto latest master although after [master bumped to play 2.8.0-M1](https://github.com/playframework/play-ws/pull/350/files#diff-0ecdbc5a001d52fb34f5eafb7cd1aaa6R27), I'm not sure if this will be available for play 2.7.

@wsargent do you plan to re-review or should I close the PR?

htmldoug commented:
Closing due to lack of review. I'll leave #267 open. I've put up a much smaller #353 to document current behavior.
